kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -156,6 +156,111 @@ static int findMaxColumnId(StructType schema) {
     return maxColumnId;
   }
 
+  static void validateColumnIds(StructType currentSchema, StructType updatedSchema) {
+    Map<Long, String> currentFieldIdToPhysicalName = new HashMap<>();
+    for (StructField field : currentSchema.fields()) {
+      validateColumnIds(field, currentFieldIdToPhysicalName);
+    }
+
+    Map<Long, String> updatedFieldIdToPhysicalName = new HashMap<>();
+    for (StructField field : updatedSchema.fields()) {
+      validateColumnIds(field, updatedFieldIdToPhysicalName);
+    }
+
+    Set<String> dedupedPhysicalNames = new HashSet<>(updatedFieldIdToPhysicalName.values());
+    if (dedupedPhysicalNames.size() != updatedFieldIdToPhysicalName.size()) {
+      throw new IllegalArgumentException("Assigned physical names must be unique");
+    }
+
+    for (Map.Entry<Long, String> field : updatedFieldIdToPhysicalName.entrySet()) {
+      String existingPhysicalName = currentFieldIdToPhysicalName.get(field.getKey());
+      // Found an existing field, verify the physical name is preserved
+      if (existingPhysicalName != null && !existingPhysicalName.equals(field.getValue())) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Existing field with id %s in current schema has physical name %s which is different from %s",
+                field.getKey(), existingPhysicalName, field.getValue()));
+      }
+    }
+  }
+
+  private static void validateColumnIds(
+      StructField field, Map<Long, String> fieldIdToPhysicalName) {
+    if (!hasPhysicalName(field)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Column mapping mode is name and field %s is missing physical name",
+              field.getName()));
+    }
+
+    if (!hasColumnId(field)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Column mapping mode is name and field %s is missing column id", field.getName()));
+    }
+
+    long columnId = getColumnId(field);
+
+    if (fieldIdToPhysicalName.containsKey(columnId)) {
+      throw new IllegalArgumentException(
+          String.format("Field %s with id %d already exists", field.getName(), columnId));
+    }
+
+    String physicalName = getPhysicalName(field);
+    fieldIdToPhysicalName.put(columnId, physicalName);
+
+    if (field.getDataType() instanceof MapType) {
+      if (!hasNestedColumnIds(field)) {
```

Comment:
Technically the Delta protoocol doesn't specify that `delta.columnMapping.nested.ids` needs to be set but I believe it's a requirement for Iceberg compat.

Reply:
I added a code comment for the method which explains that we do this check conditionally on icebergCompatV2 , let me know if that makes sense to you @allisonport-db @scottsand-db